src/ray/common/asio/periodical_runner.cc

```diff
@@ -38,102 +37,102 @@ void PeriodicalRunner::Clear() {
+          if (auto self = weak_self.lock(); self) {
+            if (RayConfig::instance().event_stats()) {
+              self->DoRunFnPeriodicallyInstrumented(
+                  std::move(fn),
+                  boost::posix_time::milliseconds(period_ms),
+                  std::move(timer),
+                  std::move(name));
+            } else {
+              self->DoRunFnPeriodically(std::move(fn),
+                                        boost::posix_time::milliseconds(period_ms),
+                                        std::move(timer));
+            }
           }
         },
         "PeriodicalRunner.RunFnPeriodically");
   }
 }
 
 void PeriodicalRunner::DoRunFnPeriodically(
-    const std::function<void()> &fn,
+    std::function<void()> fn,
     boost::posix_time::milliseconds period,
     std::shared_ptr<boost::asio::deadline_timer> timer) {
   fn();
   absl::MutexLock lock(&mutex_);
   timer->expires_from_now(period);
-  timer->async_wait(
-      [this, stopped = stopped_, fn = std::move(fn), period, timer = std::move(timer)](
-          const boost::system::error_code &error) {
-        if (*stopped) {
-          return;
-        }
-        if (error == boost::asio::error::operation_aborted) {
-          // `operation_aborted` is set when `timer` is canceled or destroyed.
-          // The Monitor lifetime may be short than the object who use it. (e.g.
-          // gcs_server)
-          return;
-        }
-        RAY_CHECK(!error) << error.message();
-        DoRunFnPeriodically(fn, period, timer);
-      });
+  auto weak_self = weak_from_this();
+  timer->async_wait([weak_self, fn = std::move(fn), period, timer = std::move(timer)](
+                        const boost::system::error_code &error) {
+    if (auto self = weak_self.lock(); self) {
+      if (error == boost::asio::error::operation_aborted) {
+        // `operation_aborted` is set when `timer` is canceled or destroyed.
+        // The Monitor lifetime may be short than the object who use it. (e.g.
+        // gcs_server)
+        return;
+      }
+      RAY_CHECK(!error) << error.message();
+      self->DoRunFnPeriodically(std::move(fn), period, std::move(timer));
+    }
+  });
 }
 
 void PeriodicalRunner::DoRunFnPeriodicallyInstrumented(
-    const std::function<void()> &fn,
+    std::function<void()> fn,
     boost::posix_time::milliseconds period,
     std::shared_ptr<boost::asio::deadline_timer> timer,
-    const std::string name) {
+    std::string name) {
   fn();
   absl::MutexLock lock(&mutex_);
   timer->expires_from_now(period);
   // NOTE: We add the timer period to the enqueue time in order only measure the time in
   // which the handler was elgible to execute on the event loop but was queued by the
   // event loop.
   auto stats_handle = io_service_.stats().RecordStart(name, period.total_nanoseconds());
-  timer->async_wait([this,
+  auto weak_self = weak_from_this();
+  timer->async_wait([weak_self,
```

Comment:
Same here, looks like `weak_self` could be moved; what about `weak_ref = weak_from_this()` in the capture list?